src/app/clusters/bindings/bindings.cpp

```diff
@@ -121,38 +99,83 @@ bool emberAfBindingClusterBindCallback(app::CommandHandler * commandObj, const a
         return true;
     }
 
+    if (BindingManager::GetInstance().CreateBinding(fabricIndex, nodeId, localEndpoint, clusterId) != CHIP_NO_ERROR)
+    {
+        emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_FAILURE);
+        return true;
+    }
     emberSetBinding(bindingIndex, &bindingEntry);
     emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_SUCCESS);
     return true;
 }
 
+uint8_t GetNumberOfBindingForNode(FabricIndex fabric, NodeId node)
+{
+    uint8_t numBinding = 0;
+    EmberBindingTableEntry entry;
+    for (uint8_t i = 0; i < EMBER_BINDING_TABLE_SIZE; i++)
+    {
+        if (emberGetBinding(i, &entry) == EMBER_SUCCESS && entry.type != EMBER_UNUSED_BINDING && entry.fabricIndex == fabric &&
+            entry.nodeId == node)
+        {
+            numBinding++;
+        }
+    }
+    return numBinding;
+}
+
 bool emberAfBindingClusterUnbindCallback(app::CommandHandler * commandObj, const app::ConcreteCommandPath & commandPath,
                                          const Commands::Unbind::DecodableType & commandData)
 {
-    auto & nodeId     = commandData.nodeId;
-    auto & groupId    = commandData.groupId;
-    auto & endpointId = commandData.endpointId;
-    auto & clusterId  = commandData.clusterId;
+    NodeId nodeId             = commandData.nodeId;
+    GroupId groupId           = commandData.groupId;
+    ClusterId clusterId       = commandData.clusterId;
+    EndpointId remoteEndpoint = commandData.endpointId;
+    EndpointId localEndpoint  = commandPath.mEndpointId;
+    FabricIndex fabricIndex   = commandObj->GetAccessingFabricIndex();
 
     ChipLogDetail(Zcl, "RX: UnbindCallback");
 
-    EmberBindingTableEntry bindingEntry;
-    if (prepareBinding(bindingEntry, nodeId, groupId, endpointId, clusterId) != EMBER_SUCCESS)
+    if (groupId && nodeId)
     {
+        ChipLogError(Zcl, "Binding: Invalid request");
         emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_MALFORMED_COMMAND);
         return true;
     }
 
+    EmberBindingTableEntry bindingEntry(fabricIndex, nodeId, groupId, localEndpoint, remoteEndpoint, clusterId);
     uint8_t bindingIndex;
     if (getBindingIndex(bindingEntry, &bindingIndex) != EMBER_SUCCESS)
     {
         emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_NOT_FOUND);
         return true;
     }
 
+    if (GetNumberOfBindingForNode(fabricIndex, nodeId) == 1)
+    {
+        CHIP_ERROR err = BindingManager::GetInstance().DisconnectDevice(fabricIndex, nodeId);
+        if (err != CHIP_NO_ERROR)
+        {
+            emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_FAILURE);
```

Comment:
Same here: log the error if not expected under normal operation path? Converting a chip error (which is supposed to  b e descriptive) to just "ZCL FAILURE" loses a lot of debuggability.